### ISSUE TITLE
MACSYMA compiler help 

### ISSUE THREAD

**Issue by TheFausap** [PDP-10/its#1023]
Hello,
I got an error using a file compiled by macsyma compiler. I'm using KLH10 and I followed the procedure done by @eswenson1 

I managed to port the diag.mac file (a very interesting piece of software - from share/contrib directory) in macsyma. Everything is fine if I use the macsyma source code.
I compiled this using:

1) translate_file("diag >");
2) compile_lisp_file("diag trlisp");

The created diag fasl has a problem with MAKELIST, using the same input data I got this error when I use the FASL file:

```
RPART FASL DSK MACSYM being loaded
Loading done
Lower bound to MAKELIST : 1, not less than upper bound: 0
```

The macsyma file load (at start) a lisp file (diagl lisp) with some utility functions. Macsyma compiler complains about those undefined functions, but the support file (diagl lisp) is loaded when I do a load("diag"), so they should be ok, or not ? 

Is this the correct method to compile multiple files in Macsyma ?

**Comment by eswenson1** [comment]
Can you provide your DIAG > file?

**Comment by TheFausap** [comment]
sure. here there are the two files:

https://gist.github.com/TheFausap/5c96f1594c89ded22f33004ff1baef3b.js

**Comment by eswenson1** [comment]
I have a different issue. When I try to load DIAG FASL (compiled with translate_file and compile_lisp_file), I get this error:

> (C5) load("diag");
> DIAG FASL DSK EJS1 being loaded
> Could not find file which matches "DIAGL *"

Am I missing a file I need?

**Comment by eswenson1** [comment]
Sorry, I see you did post that file.  Let me try again.

**Comment by eswenson1** [comment]
Hmm... I don't see the issue.  When I have DIAGL LISP in my home directory, and then do a 'load("diag");' from macsyma, I get this:

> *macsym↑K!
> 
> This is MACSYMA 1000
> 
> (C1) load("ejs1\;diag");
> 
> DIAG FASL DSK EJS1 being loaded
> DIAGL LISP DSK EJS1 being loaded
> Loading done
> Loading done
> (D1)                                                  "DSK:EJS1;DIAG FASL"
> 
> (C2)

**Comment by eswenson1** [comment]
Given that this is loaded in my Macsyma, what I type to see if it actually works?  (I don't remember much Macsyma and nothing about matrices).

**Comment by eswenson1** [comment]
Ah, I get your error when I do this:
> (C2) diag([1,2,3]);
> 
> ACALL FASL DSK MACSYM being loaded
> Loading done
> Lower bound to MAKELIST : 1, not less than upper bound: 0
> 
> (C3)

It's not from RPART (in my case), but it is the same error.

**Comment by eswenson1** [comment]
The call stack looks like this at the time of the error:

>(APPLY #'DEBUG-PARSE-ALL-FRAMES
       'NIL)
 (DEBUG)
  (APPLY #'*BREAK
         '(T "LISP  (Type <Alt>P<Space> to continue.)"))
   (MPAUSE NIL)
    (APPLY #'(LAMBDA (NIL NIL) (MPAUSE NIL))
           '(#FILE-IN-|TTY:* *|-70012 2))
     (APPLY #'*TYI
            'NIL)
      (APPLY #'RETRIEVE1
             '(NIL BREAK))
       (APPLY #'ERRBREAK
              '(NIL))
        (APPLY #'ERROR
               'NIL)
         (APPLY #'*MERROR
                '(((# ACALL /71) . 1) $MAKELIST 1 0))
          (APPLY #'$DIAG
                 '(((MLIST SIMP) 1 2 3)))
           (APPLY #'*APPLY
                  '($DIAG ((# 1 2 3))))
            (APPLY #'MEVAL1
                   '((($DIAG) (# 1 2 3))))
             (APPLY #'MEVAL
                    '((($DIAG) (# 1 2 3))))
              (APPLY #'MEVAL*
                     '((($DIAG) (# 1 2 3))))
               (APPLY #'CONTINUE1
                      '((($DIAG) (# 1 2 3))))
                (CONTINUE)
                 (APPLY #'EVAL
                        '(0))
                  (APPLY #'MAPC
                         '(((SUPUNBIND) (CONTINUE)) ((CONTINUE))))

**Comment by eswenson1** [comment]
The error we're seeing comes from TRANS; ACALL 71:

> (DEFMFUN INTERVAL-ERROR (FUN LOW HIGH)
>  (MERROR "Lower bound to ~:@M : ~M, not less than upper bound: ~M"
>	  FUN LOW HIGH))

**Comment by eswenson1** [comment]
Are you sure the logic in DIAG MAC isn't flawed?  It is calling MAKELIST and it is that call that is failing.  It seems that the bounds passed in as the 3rd and 4th parameter are not such that the low bound is lower than the high bound.

**Comment by TheFausap** [comment]
@eswenson1 if you try to load the mac file, instead of the compiled one, everything is fine.
That's the issue, I think the problem is with SECOND. This function doesn't exist in MACSYMA (only in MAXIMA), so I added it in DIAGL.LISP, but I suppose when it's used in the FASL something bad happens.

**Comment by TheFausap** [comment]
![diag1](https://user-images.githubusercontent.com/1846443/41880184-0c7c34e4-78dd-11e8-8b6c-f29da6a768d3.PNG)

**Comment by eswenson1** [comment]
Yeah, just got that too!

**Comment by eswenson1** [comment]
So the bug is with the translator.  If you translate DIAG > to DIAG TRLISP and load that, it loads fine, but when you run diag([1,2,3]) you get the MAKELIST error.

**Comment by eswenson1** [comment]
In other words, I didn't use the DIAG FASL file at all and still saw the issue.

**Comment by eswenson1** [comment]
And here is the code for the translated DIAG function:

> (DEFMTRFUN ($DIAG $ANY MDEFINE NIL NIL) ($LST) NIL ((LAMBDA ($LST $SIZE $LEFT $ROWS) NIL (DO (($A) (MDO (CDR $LST) (CDR MDO))) ((N!
ULL MDO) (QUOTE $DONE)) (SETQ $A (CAR MDO)) (SETQ $SIZE (ADD* $SIZE ($LENGTH $A)))) (DO (($A) (MDO (CDR $LST) (CDR MDO))) ((NULL M!
DO) (QUOTE $DONE)) (SETQ $A (CAR MDO)) (DO (($R) (MDO (CDR (SIMPLIFY ($ARGS $A))) (CDR MDO))) ((NULL MDO) (QUOTE $DONE)) (SETQ $R !
(CAR MDO)) (SETQ $ROWS ($CONS ($APPEND ((LAMBDA (G0017 G0018) (COND ((NOT (< G0018 G0017)) (DO (($I G0017 (1+ $I)) (G0019 NIL (CON!
S 0. G0019))) ((> $I G0018) (LIST* (QUOTE (MLIST)) (NREVERSE G0019))) (DECLARE (FIXNUM $I)))) (T (INTERVAL-ERROR (QUOTE $MAKELIST)!
 G0017 G0018)))) 1. $LEFT) $R ((LAMBDA (G0020 G0021) (COND ((NOT (< G0021 G0020)) (DO (($I G0020 (1+ $I)) (G0022 NIL (CONS 0. G002!
2))) ((> $I G0021) (LIST* (QUOTE (MLIST)) (NREVERSE G0022))) (DECLARE (FIXNUM $I)))) (T (INTERVAL-ERROR (QUOTE $MAKELIST) G0020 G0!
021)))) 1. (ADD* $SIZE (*MMINUS (ADD* $LEFT ($LENGTH $R)))))) $ROWS))) (SETQ $LEFT (ADD* $LEFT ($LENGTH (SIMPLIFY ($FIRST $A))))))!
 (SIMPLIFY (MAPPLY-TR (TRD-MSYMEVAL $MATRIX (QUOTE $MATRIX)) ($REVERSE $ROWS)))) (SIMPLIFY (MAP1 (GETOPR (M-TLAMBDA ($A) NIL (COND!
 (($MATRIXP $A) $A) (T (SIMPLIFY ($MATRIX (LIST (QUOTE (MLIST)) $A))))))) $LST)) 0. 0. (QUOTE ((MLIST)))))

**Comment by TheFausap** [comment]
Maybe MAKELIST in LISP is more restrictive than in MACSYMA.
If you run makelist(a,i,3,1); you have a different error, but if you run makelist(a,i,1,0); you have [ ] and not an error.

**Comment by eswenson1** [comment]
And the error is being raised directly by that code.  I ran diag([1,2,3]); with DIAG TRLISP loaded (no FASL, no MAC) and with INTERVAL-ERROR being traced, and got this:

> (C3) diag([1,2,3]);
>
>(1 ENTER INTERVAL-ERROR ($MAKELIST 1 0))
>Lower bound to MAKELIST : 1, not less than upper bound: 0
>
> (C4)

So the translated code (see previous comment) is explicitly calling this error.

The MAKELIST being called here is the macsyma MAKELIST ($MAKELIST), not the lisp MAKELISP.

**Comment by eswenson1** [comment]
I think the issue is this:  we *are* calling the same MAKELIST -- the macsyma one.  However, in the case of running DIAG MAC, we autoload $MAKELIST from MAXOUT; MAKEL FASL.  However, when running either DIAG TRLISP or DIAG FASL, the translator has inlined the call to $MAKELIST.  As you can see from the DIAG TRLISP code, we are not actually calling $MAKELIST.  We have this definition inlined (twice) in the definition of $DIAG:

> (LAMBDA (G0017 G0018)
                                                (COND ((NOT (< G0018 G0017))
                                                       (DO (($I G0017 (1+ $I))
                                                            (G0019 NIL (CONS 0. G0019)))
                                                           ((> $I G0018)
                                                            (LIST* (QUOTE (MLIST)) (NREVERSE G0019)))
                                                           (DECLARE (FIXNUM $I))))
                                                      (T (INTERVAL-ERROR (QUOTE $MAKELIST) G0017 G0018)))) 

And it is this definition that decides to take the "T" branch of the COND -- because it has computed that  (< G0018 G0017) is true.  So I guess the question is why is G0018 less than G0017 in the translated code?

**Comment by eswenson1** [comment]
And, as you said earlier, why when you call makelist(0,i,1,2) and makelist(0,i,2,1) do both of these do something sensible and not raise an error -- when the MAKEL FASL version of makelist is used.

**Comment by eswenson1** [comment]
I created a test macsyma source with the following contents:

> testdiag (lst) := makelist(0,i,0,-1);

And I translated this to LISP and got this:

> (DEFMTRFUN ($TESTDIAG $ANY MDEFINE NIL NIL) ($LST) NIL ((LAMBDA (G0017 G0018) (COND ((NOT (< G0018 G0017)) (DO (($I G0017 (1+ $I))!
 (G0019 NIL (CONS 0. G0019))) ((> $I G0018) (LIST* (QUOTE (MLIST)) (NREVERSE G0019))) (DECLARE (FIXNUM $I)))) (T (INTERVAL-ERROR (!
QUOTE $MAKELIST) G0017 G0018)))) 0. -1.))

And when I loaded and ran (with testdiag([1,2,3]);) the translated version, I got the same error we've been seeing:

> (C4) testdiag([1,2,3]);
> Lower bound to MAKELIST : 0, not less than upper bound: - 1

> (C5) makelist(0,i,0,-1);

> MAKEL FASL DSK MAXOUT being loaded
> Loading done
> (D5)                                                           []

So testdiag calls makelist(0,i,0,-1) and gets the bounds error (which is actually a reasonable error). However, calling makelist(0,i,0,-1) directly from Macsyma, which loads the MAKEL FASL implementation of $makelist, you don't get any error.  

For some reason, when the real DIAG is translated, it is passing 0, -1 as the 3rd and 4th arguments to $MAKELIST, but when it is run from source, this isn't happening.

**Comment by eswenson1** [comment]
I found the code in the macsyma translator (TRANSL;TRANS1 108) where $MAKELIST is being translated.  I can map the code in that file to the code in the DIAG TRLISP that is generated -- at least by eyeballing.  I wonder if this translate logic is flawed.  I wonder if we have the latest version of TRANS1. 

I just checked and TRANS1 108 is the latest version that ever appeared on ITS backup tapes. And it was last modified in 1982.

**Comment by eswenson1** [comment]
Looking over the translator (TRANSL; TRANS1 108) for $MAKELIST, I don't see anything wrong with it.  Looking over the lisp implementation of $MAKELIST in DAS; MSTUFF 21, it doesn't look like it enforces the fact that the 3rd argument should be less than the 4th argument.  So I think things may be working as they were "intended".  When you translate the code for DIAG, the translated $MAKELIST calls are more strict -- they enforce the 3rd and 4th argument restriction.  When you run it as macsyma code, the compiled $MAKELIST code doesn't seem to check/care about this restriction.  

Can you, @TheFausap, who understands more about all this math stuff (like what DIAG is doing) say for sure that the implementation isn't just "illegal" and relying on calling $MAKELIST incorrectly?

**Comment by eswenson1** [comment]
Looking over this definition of diag again (from diag mac), we see this:

> diag (lst) :=
  block([lst:map(lambda([A], if matrixp(A) then A else matrix([A])),lst),
	 size:0,left:0,rows:[]],
  for A in lst do size : size + length(A),
  for A in lst do (
    for r in args(A) do
      rows : cons(append(makelist(0,i,1,left), r,
			 makelist(0,i,1,size-(left + length(r)))),
			 rows),
      left : left + length(first(A))),
    apply(matrix,reverse(rows)))$

It seems to me that left is set to be "left + length(first(A))" AFTER it is used in the call the makelist.  left is initialized to 0 on function entry.  So it seems wrong. 

I changed the definition of diag to this:

> mymakelist(a,b,c,d) :=
>  if is(d = 0) then [] else makelist(a,b,c,d)$
>
>diag (lst) :=
  block([lst:map(lambda([A], if matrixp(A) then A else matrix([A])),lst),
     size:0,left:0,rows:[]],
  for A in lst do size : size + length(A),
  for A in lst do (
    for r in args(A) do
      rows : cons(append(mymakelist(0,i,1,left), r,
             mymakelist(0,i,1,size-(left + length(r)))),
             rows),
    left : left + length(first(A))),
    apply(matrix,reverse(rows)))$

And now, regardless of whether I run the macsyma source, the translated lisp, or the compiled fasl, it works as expected.

Conclusion: the code in diag was incorrectly calling makelist and relying on the fact that it didn't check the pre-condition that it claims to (namely that the high bound was higher than the low bound).  The macsyma translator for makelist was enforcing these pre-conditions.

**Comment by TheFausap** [comment]
Yes! In Maxima latest version, diag is loaded using the source maxima file, not a compiled one :)
